### DISCUSSION TITLE
Combined interfaces using a using statement

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#6422] [Language Ideas] [answered]
Sometimes you want to restrict objects to those that implement two or more interfaces, and there isn't really a clean way to do it; you have to make a dummy interface which implements all the interfaces:

```
public interface IMobileLegged : IMobile, ILegged
{
}
```

Which of course won't even work if you can't modify the types that implement the interfaces (unless we get duck typed interfaces or mixins or whatever you want to call them, but that's another story).

So, how about this? We could use a `using` statement and an operator (perhaps `+` or `|`) to combine interfaces together on the fly!

```
using IMobileLegged = IMobile + ILegged;

public static void Walk(this IMobileLegged obj)
{
    //  here we call members of IMobile and ILegged
}

public static void WalkGeneric<T>(this T obj) where T : IMobileLegged
{
    // or do it this way
}
```

**Comment by HaloFour** [comment]
The syntax isn't the problem as much as the metadata is.  The runtime has no concept of a `IMobile + ILegged` type and the method signature requires a type.

**Reply by ekolis** [reply]
Oh, true. I suppose implementing this would basically require implementing mixins?

**Comment by PathogenDavid** [comment]
Why not just use generic constraints? For example, your `WalkGeneric` could just be

```csharp
public static void WalkGeneric<T>(this T obj) where T : IMobile, ILegged
{
    // or do it this way
}
```

**Reply by ekolis** [reply]
That could work!

**Comment by svick** [comment, marked as answer]
This is a subset of https://github.com/dotnet/csharplang/discussions/399.
